File: frontend/lib/soc.ts
import { apiRequest } from "./api";

export type Severity = "Low" | "Medium" | "High" | "Critical";

export type SocAnalyzePayload = {
  title: string;
  log_content: string;
  source_ip: string;
  destination_ip: string;
  username: string;
  event_type: string;
};

export type MitreTechnique = {
  technique_id: string;
  name: string;
  tactic: string;
  description: string;
  common_log_sources: string[];
  detection_ideas: string[];
  mitigation: string[];
  example_keywords: string[];
  matched_keywords: string[];
};

export type SocAnalysisResult = {
  severity: Severity;
  risk_score: number;
  summary: string;
  suspicious_indicators: string[];
  mitre_techniques: MitreTechnique[];
  recommended_actions: string[];
  incident_report: string;
};

export type InvestigationHistoryItem = {
  id: number;
  title: string;
  source_ip: string;
  destination_ip: string;
  username: string;
  event_type: string;
  severity: Severity;
  risk_score: number;
  summary: string;
  mitre_mappings_json?: string | null;
  created_at: string;
};

export async function analyzeSocAlert(payload: SocAnalyzePayload, token: string) {
  return apiRequest<SocAnalysisResult>(
    "/api/soc/analyze",
    {
      method: "POST",
      body: JSON.stringify(payload),
    },
    token,
  );
}

export async function fetchSocHistory(token: string) {
  return apiRequest<InvestigationHistoryItem[]>("/api/soc/history", { method: "GET" }, token);
}

File: backend/app/schemas/soc.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

Severity = Literal["Low", "Medium", "High", "Critical"]


class AlertAnalysisRequest(BaseModel):
    title: str = Field(..., min_length=3, max_length=255)
    log_content: str = Field(..., min_length=5)
    source_ip: str = Field(..., min_length=2, max_length=64)
    destination_ip: str = Field(..., min_length=2, max_length=64)
    username: str = Field(..., min_length=1, max_length=120)
    event_type: str = Field(..., min_length=2, max_length=80)


class MitreTechnique(BaseModel):
    technique_id: str
    name: str
    tactic: str
    description: str
    common_log_sources: list[str]
    detection_ideas: list[str]
    mitigation: list[str]
    example_keywords: list[str]
    matched_keywords: list[str] = []


class AlertAnalysisResponse(BaseModel):
    severity: Severity
    risk_score: int = Field(..., ge=0, le=100)
    summary: str
    suspicious_indicators: list[str]
    mitre_techniques: list[MitreTechnique]
    recommended_actions: list[str]
    incident_report: str


class MitreLogMappingRequest(BaseModel):
    log_content: str = Field(..., min_length=5)
    event_type: str = Field(default="", max_length=80)


class InvestigationRead(BaseModel):
    id: int
    title: str
    source_ip: str
    destination_ip: str
    username: str
    event_type: str
    severity: Severity
    risk_score: int
    summary: str
    mitre_mappings_json: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}
